### ISSUE TITLE
Nested content within nested content is not being escaped

### ISSUE THREAD

**Issue by Frost117** [Jumoo/Jumoo.TranslationManager.Issues#15]
First thing first, Brilliant package you have here Kevin. It does wonderful stuff with the Azure cognitive service.

We have a client who has been using this package for all of their translation needs, where it works beautifully on the source env.
When they transfer content node that has nested content within nested content, issues begins to come up.

The nested content within the nested content, comes up as null when you transfer the translated nodes over with Umbraco Deploy.

I think I have the reason why. When you manually created a content node and manually translate it, it will look like this:

`[
    {
        "name":"Statement",
        "ncContentTypeAlias":"statement",
        "key":"d09aff6e-3a2e-4ccb-95f5-84fd78e46a48",
        "title":"Manual title",
        "titleTag":"[\"h1\"]",
        "body":"<p>Her er bodytekst med manualt arbejde involveret</p>",
        "buttons":"
        [{
                \"name\":\"Item 1\",
                \"ncContentTypeAlias\":\"button\",
                \"key\":\"bc13bb58-2846-471f-a8a9-a50ca6c03799\",
                \"link\":\"umb://document/9d08af7975d64c2c9f06d47ffee58cb4\",
                \"variant\":\"[\\\"Primary\\\"]\",
                \"buttonText\":\"adsad\"
        }]"
    }
]`

When the translation manager creates and publish a node, it will look like this:

`[
  {
    "key": "2e0f1ce4-39c0-403a-baf2-1b0743687ae2",
    "name": "Statement",
    "ncContentTypeAlias": "statement",
    "title": "Prøv at fortælle vittigheder",
    "titleTag": "[\"h1\"]",
    "body": "<p>Gør sjove noices</p>",
    "buttons": 
    [{
        "key": "c2a47107-c858-446a-8b41-7d3f89d78978",
        "name": "Item 1",
        "ncContentTypeAlias": "button",
        "link": "umb://document/9d08af7975d64c2c9f06d47ffee58cb4",
        "variant": "[\"Primary\"]",
        "buttonText": "Nå ja, ser hvem der er inde igen"
    }]
  }
]`

The manually created node contains a JSON string for buttons, so it needs to escape quotes (because it's already enclosed in quotes), but the translated node contains a JSON object. I assume the first is correct, so it's something in the translation code that doesn't correctly write the value back to the content node.

The translation does get transferred over with no issue in the transfer process. The only kink is that when you view the content in the backoffice you will see this node with empty nested content:

![image](https://user-images.githubusercontent.com/32164310/129719074-21f68933-b3da-4d68-b7a9-5fb85fd9fb50.png)

In the database, the "buttons" data will be listed as null.

Here is the node in its origin language:
![image](https://user-images.githubusercontent.com/32164310/129719164-ef590c23-c503-463f-9691-62498be7def2.png)

**Comment by KevinJump** [comment]
Thanks for the in-depth analysis, we will take a look . i suspect we need to escape all nested properties but we have to make sure it doesn't break something deeper down the tree (e.g nested in nested in nested 😄 )

**Comment by KevinJump** [comment]
Hi, 

Can you just confirm what version of Umbraco & Translation Manager you have ?

I thought i new the issue, so changed the code (and it all still works!) but when i was doing my final checks on a fresh Umbraco install i noticed that it doesn't appear that Nested, Nested content is stored escaped like in your example (before its gone through translation manager). 

So. as an example created in Umbraco I have this stored in the DB. 

```
[
  {
    "key": "c570bc18-9863-4ac4-a8c2-e0c14d809aa4",
    "name": "Title",
    "ncContentTypeAlias": "nestedNested",
    "title": "Title",
    "nested": [
      {
        "key": "0af5da01-06cf-4cd7-b18d-c58d91b63ff0",
        "name": "Feature Name",
        "ncContentTypeAlias": "feature",
        "featureName": "Feature Name",
        "featureDetails": "Feature details "
      },
      {
        "key": "72f38732-4f57-4e67-8df5-f808efcf4f2e",
        "name": "Nested Nested Title",
        "ncContentTypeAlias": "nestedNested",
        "title": "Nested Nested Title",
        "nested": [
          {
            "key": "bcc9ca64-0c72-4cf0-940f-111154f9b41a",
            "name": "Nested Feature",
            "ncContentTypeAlias": "feature",
            "featureName": "Nested Feature",
            "featureDetails": "Details, details, details."
          }
        ]
      }
    ]
  }
]
```
_(data from the umbracoPropertyData table TextValue column)_

So its not escaped in any way . _equally the translated content is also in this format, not escaped_ 

Is your nested content element inside anything else (e.g in a grid or anything ?) are there any special properties in there i might have missed (I've just being doing text/rich-text, nested in nested etc)

**It is possible Umbraco Deploy is doing something, and there is a thing in here that we are missing - but the raw DB values are stored as JSON all the way down.**

**Comment by KevinJump** [comment]
For compleatness i added a few drop downs. 

```
[
  {
    "key": "c570bc18-9863-4ac4-a8c2-e0c14d809aa4",
    "name": "Title",
    "ncContentTypeAlias": "nestedNested",
    "title": "Title",
    "nested": [
      {
        "key": "0af5da01-06cf-4cd7-b18d-c58d91b63ff0",
        "name": "Updated",
        "ncContentTypeAlias": "feature",
        "heading": [
          "h1"
        ],
        "featureName": "Updated",
        "variant": [
          "Primary"
        ],
        "featureDetails": "Feature details "
      },
      {
        "key": "72f38732-4f57-4e67-8df5-f808efcf4f2e",
        "name": "Nested Nested Title",
        "ncContentTypeAlias": "nestedNested",
        "title": "Nested Nested Title",
        "nested": [
          {
            "key": "bcc9ca64-0c72-4cf0-940f-111154f9b41a",
            "name": "Nested Feature",
            "ncContentTypeAlias": "feature",
            "heading": [
              "h2"
            ],
            "featureName": "Nested Feature",
            "variant": [
              "Secondary"
            ],
            "featureDetails": "Details, details, details."
          }
        ]
      }
    ]
  }
]
```

**Comment by KevinJump** [comment]
i can say in Umbraco v8.4.2 - the text is encoded : 

```
[{"key":"60f5ed96-a9c8-43d5-964d-d6c7143bd78b","name":"Some content","ncContentTypeAlias":"nestedWithNested","title":"Some content","body":"<p>Some content <strong>body</strong></p>","nested":"[{\"key\":\"2e6a0bbd-0d59-4e0a-a1d8-6b771dfa8137\",\"name\":\"Nested Content Item\",\"ncContentTypeAlias\":\"nestedContentItem\",\"title\":\"Nested Content Item\",\"words\":\"<p>Nested content</p>\",\"list\":\"One\\r\\nTwo\\r\\nThree\"},{\"key\":\"7bd2bae4-3481-4f04-a689-d07a8d8bb237\",\"name\":\"Double Nested Content\",\"ncContentTypeAlias\":\"nestedWithNested\",\"title\":\"Double Nested Content\",\"body\":\"<p>Some formatted <strong>content</strong> in here</p>\",\"nested\":\"[{\\\"key\\\":\\\"6c00527e-e874-46d8-ba1a-fb6f8cd18a31\\\",\\\"name\\\":\\\"More nested content\\\",\\\"ncContentTypeAlias\\\":\\\"nestedContentItem\\\",\\\"title\\\":\\\"More nested content\\\",\\\"words\\\":\\\"<p>More <a href=\\\\\\\"/{localLink:umb://document/d7c6fb8931d44ed9af6378b077ac2255}\\\\\\\" title=\\\\\\\"Child Content\\\\\\\">formatting</a></p>\\\",\\\"list\\\":\\\"List One\\\\r\\\\nList Two\\\"}]\"}]"}]
```

in Umbraco 8.15.1 it is stored as a big blob of JSON. 

```
[
  {
    "key": "c570bc18-9863-4ac4-a8c2-e0c14d809aa4",
    "name": "Title",
    "ncContentTypeAlias": "nestedNested",
    "title": "Title",
    "nested": [
      {
        "key": "0af5da01-06cf-4cd7-b18d-c58d91b63ff0",
        "name": "Updated",
        "ncContentTypeAlias": "feature",
        "heading": [
          "h1"
        ],
        "featureName": "Updated",
        "variant": [
          "Primary"
        ],
        "featureDetails": "Feature details "
      },
      {
        "key": "72f38732-4f57-4e67-8df5-f808efcf4f2e",
        "name": "Nested Nested Title",
        "ncContentTypeAlias": "nestedNested",
        "title": "Nested Nested Title",
        "nested": [
          {
            "key": "bcc9ca64-0c72-4cf0-940f-111154f9b41a",
            "name": "Nested Feature",
            "ncContentTypeAlias": "feature",
            "heading": [
              "h2"
            ],
            "featureName": "Nested Feature",
            "variant": [
              "Secondary"
            ],
            "featureDetails": "Details, details, details."
          }
        ]
      }
    ]
  }
]
```

going to see if I can identify when that happened - and make sure we work both ways.

**Comment by Frost117** [comment]
Absolutely Kevin.

I am using 8.14.3 with v8.7.3.0 of the translation manager

Is your nested content element inside anything else (e.g in a grid or anything ?) 
are there any special properties in there i might have missed (I've just being doing text/rich-text, nested in nested etc)

Current setup with the nested content property looks like this:
![image](https://user-images.githubusercontent.com/32164310/129873090-b667123d-a406-4caa-b0ef-0900207b6e00.png)

Nested content datatype
![image](https://user-images.githubusercontent.com/32164310/129873396-21559cec-fe8f-4718-a30e-8f9ecc76e753.png)

The Statement datatype is the 2nd nested content datatype, that has the following:
textstring
dropdown
content picker

**Comment by KevinJump** [comment]
I am going to do some release testing but after some more investigation i think. we can escape it and later versions of Umbraco will cope when they turn it back into the json that is stored.

**Comment by KevinJump** [comment]
Hi, 

just released v8.7.4 - which i **think** fixes this issue on all versions of Umbraco 8 🤞

https://www.nuget.org/packages/Jumoo.TranslationManager/8.7.4
